db.go

```diff
@@ -556,20 +556,117 @@ func (db *DB) reload(deleteable ...string) (err error) {
 	return errors.Wrap(db.head.Truncate(maxt), "head truncate failed")
 }
 
+// ValidateBlockSequence returns error if given block meta files indicate that some blocks overlaps within sequence.
 func validateBlockSequence(bs []*Block) error {
-	if len(bs) == 0 {
+	if len(bs) <= 1 {
+		return nil
+	}
+
+	var metas []BlockMeta
+	for _, b := range bs {
+		metas = append(metas, b.meta)
+	}
+
+	overlaps := OverlappingBlocks(metas)
+	if len(overlaps) > 0 {
+		return errors.Errorf("block time ranges overlap: %s", SprintOverlappedBlocks(overlaps))
+	}
+
+	return nil
+}
+
+type TimeRange struct {
+	Min, Max int64
+}
+
+// OverlappingBlocks returns all overlapping blocks from given meta files aggregated by overlaping range.
+// We sort blocks by minTime. Then we iterate over each block minTime and treat it as our "current" timestamp.
```

Comment:
That line onwards belong inside of the function. People reading the `godoc` do not care how it works internally.